File: crypto/crypto-common/src/main/java/org/openecard/crypto/common/sal/did/DidInfo.java
/****************************************************************************
 * Copyright (C) 2016 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/

package org.openecard.crypto.common.sal.did;

import iso.std.iso_iec._24727.tech.schema.ACLList;
import iso.std.iso_iec._24727.tech.schema.ACLListResponse;
import iso.std.iso_iec._24727.tech.schema.AccessControlListType;
import iso.std.iso_iec._24727.tech.schema.CertificateRefType;
import iso.std.iso_iec._24727.tech.schema.DIDAbstractMarkerType;
import iso.std.iso_iec._24727.tech.schema.DIDAuthenticate;
import iso.std.iso_iec._24727.tech.schema.DIDAuthenticateResponse;
import iso.std.iso_iec._24727.tech.schema.DIDGet;
import iso.std.iso_iec._24727.tech.schema.DIDGetResponse;
import iso.std.iso_iec._24727.tech.schema.DIDScopeType;
import iso.std.iso_iec._24727.tech.schema.DIDStructureType;
import iso.std.iso_iec._24727.tech.schema.Hash;
import iso.std.iso_iec._24727.tech.schema.HashResponse;
import iso.std.iso_iec._24727.tech.schema.PinCompareDIDAuthenticateInputType;
import iso.std.iso_iec._24727.tech.schema.Sign;
import iso.std.iso_iec._24727.tech.schema.SignResponse;
import iso.std.iso_iec._24727.tech.schema.TargetNameType;
import java.io.ByteArrayInputStream;
import java.math.BigInteger;
import java.security.cert.Certificate;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import javax.annotation.Nullable;
import javax.xml.parsers.ParserConfigurationException;
import org.openecard.common.ECardConstants;
import org.openecard.common.SecurityConditionUnsatisfiable;
import org.openecard.common.WSHelper;
import org.openecard.common.anytype.pin.PINCompareDIDAuthenticateInputType;
import org.openecard.common.anytype.pin.PINCompareDIDAuthenticateOutputType;
import org.openecard.common.util.ByteUtils;
import org.openecard.crypto.common.sal.ACLResolver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 *
 * @author Tobias Wich
 */
public class DidInfo {

    private static final Logger LOG = LoggerFactory.getLogger(DidInfo.class);

    private final DidInfos didInfos;
    private final byte[] application;
    private final TargetNameType didTarget;
    @Nullable
    private DIDScopeType didScope;
    @Nullable
    private char[] pin;

    private CertificateFactory certFactory;
    private List<X509Certificate> certs;

    DidInfo(DidInfos didInfos, byte[] application, String didName, @Nullable char[] pin) {
	this.didInfos = didInfos;
	this.application = ByteUtils.clone(application);
	this.didTarget = new TargetNameType();
	this.didTarget.setDIDName(didName);
	if (pin != null) {
	    Arrays.fill(this.pin, ' ');
	    this.pin = pin.clone();
	} else {
	    this.pin = null;
	}

	// we are in the same application, so set it to null and get the real value
	this.didScope = null;
	try {
	    DIDStructureType didStruct = getDID();
	    if (didStruct != null) {
		this.didScope = didStruct.getDIDScope();
	    }
	} catch (WSHelper.WSException ex) {
	    // too bad ;-)
	    LOG.warn("Failed to retrieve DID qualifier.");
	}
    }

    DidInfo(DidInfos didInfos, byte[] application, DIDStructureType didStruct, @Nullable char[] pin) {
	this.didInfos = didInfos;
	this.application = ByteUtils.clone(application);
	this.didTarget = new TargetNameType();
	this.didTarget.setDIDName(didStruct.getDIDName());
	this.didScope = didStruct.getDIDScope();
	if (pin != null) {
	    Arrays.fill(this.pin, ' ');
	    this.pin = pin.clone();
	} else {
	    this.pin = null;
	}
    }

    public void setPin(char[] pin) {
	if (pin != null) {
	    Arrays.fill(this.pin, ' ');
	    this.pin = pin.clone();
	} else {
	    this.pin = null;
	}
    }

    public AccessControlListType getACL() throws WSHelper.WSException {
	ACLList req = new ACLList();
	req.setConnectionHandle(didInfos.getHandle(application));
	req.setTargetName(didTarget);

	ACLListResponse res = (ACLListResponse) didInfos.getDispatcher().safeDeliver(req);
	WSHelper.checkResult(res);

	return res.getTargetACL();
    }

    public DIDStructureType getDID() throws WSHelper.WSException {
	DIDGet req = new DIDGet();
	req.setConnectionHandle(didInfos.getHandle(application));
	req.setDIDName(getDidName());
	req.setDIDScope(didScope);

	DIDGetResponse res = (DIDGetResponse) didInfos.getDispatcher().safeDeliver(req);
	WSHelper.checkResult(res);

	return res.getDIDStructure();
    }

    public String getDidName() {
	return didTarget.getDIDName();
    }

    protected synchronized CertificateFactory getCertFactory() throws CertificateException {
	if (certFactory == null) {
	    certFactory = CertificateFactory.getInstance("X.509");
	}
	return certFactory;
    }

    public String getProtocol() throws WSHelper.WSException {
	return getDID().getDIDMarker().getProtocol();
    }

    public boolean isAuthenticated() throws WSHelper.WSException {
	return getDID().isAuthenticated();
    }

    public List<DIDStructureType> getMissingDids() throws WSHelper.WSException, SecurityConditionUnsatisfiable {
	ACLResolver resolver = new ACLResolver(didInfos.getDispatcher(), didInfos.getHandle(application));
	List<DIDStructureType> missingDids = resolver.getUnsatisfiedDIDs(didTarget, getACL().getAccessRule());
	return missingDids;
    }

    public List<DidInfo> getMissingDidInfos() throws WSHelper.WSException, SecurityConditionUnsatisfiable, NoSuchDid {
	ArrayList<DidInfo> result = new ArrayList<>();
	for (DIDStructureType didStruct : getMissingDids()) {
	    result.add(didInfos.getDidInfo(didStruct.getDIDName()));
	}
	return result;
    }

    public boolean isPinSufficient() throws WSHelper.WSException {
	try {
	    List<DIDStructureType> missingDids = getMissingDids();
	    // check if there is anything other than a pin did in the list
	    for (DIDStructureType missingDid : missingDids) {
		DidInfo infoObj;
		if (missingDid.getDIDScope() == DIDScopeType.GLOBAL) {
		    infoObj = didInfos.getDidInfo(missingDid.getDIDName());
		} else {
		    infoObj = didInfos.getDidInfo(application, missingDid.getDIDName());
		}
		if (! infoObj.isPinDid()) {
		    return false;
		}
	    }
	    // no PIN DID in missing list
	    return true;
	} catch (SecurityConditionUnsatisfiable ex) {
	    // not satisfiable means pin does not suffice
	    return false;
	} catch (NoSuchDid ex) {
	    String msg = "DID referenced in CIF could not be resolved.";
	    LOG.error(msg, ex);
	    throw WSHelper.createException(WSHelper.makeResultError(ECardConstants.Minor.App.INT_ERROR, msg));
	}
    }

    public boolean needsPin() throws WSHelper.WSException, SecurityConditionUnsatisfiable {
	try {
	    List<DIDStructureType> missingDids = getMissingDids();
	    // check if there is a pin did in the list
	    for (DIDStructureType missingDid : missingDids) {
		DidInfo infoObj;
		if (missingDid.getDIDScope() == DIDScopeType.GLOBAL) {
		    infoObj = didInfos.getDidInfo(missingDid.getDIDName());
		} else {
		    infoObj = didInfos.getDidInfo(application, missingDid.getDIDName());
		}
		if (infoObj.isPinDid()) {
		    return true;
		}
	    }
	    // no PIN DID in missing list
	    return false;
	} catch (NoSuchDid ex) {
	    String msg = "DID referenced in CIF could not be resolved.";
	    LOG.error(msg, ex);
	    throw WSHelper.createException(WSHelper.makeResultError(ECardConstants.Minor.App.INT_ERROR, msg));
	}
    }


    public DIDAbstractMarkerType getRawMarker() throws WSHelper.WSException {
	return getDID().getDIDMarker();
    }


    public boolean isPinDid() throws WSHelper.WSException {
	return "urn:oid:1.3.162.15480.3.0.9".equals(getProtocol());
    }

    public PinCompareMarkerType getPinCompareMarker() throws WSHelper.WSException {
	return new PinCompareMarkerType(getDID().getDIDMarker());
    }

    @Nullable
    public BigInteger enterPin(@Nullable char[] pin) throws WSHelper.WSException {
	if (! isPinDid()) {
	    throw new IllegalStateException("Enter PIN called for a DID which is not a PIN DID.");
	}

	try {
	    PinCompareDIDAuthenticateInputType data = new PinCompareDIDAuthenticateInputType();
	    data.setProtocol(getProtocol());
	    // add PIN
	    if (pin != null && pin.length != 0) {
		PINCompareDIDAuthenticateInputType builder = new PINCompareDIDAuthenticateInputType(data);
		builder.setPIN(pin);
		data = builder.getAuthDataType();
		builder.setPIN(null);
	    }

	    DIDAuthenticate req = new DIDAuthenticate();
	    req.setConnectionHandle(didInfos.getHandle(application));
	    req.setDIDName(didTarget.getDIDName());
	    req.setDIDScope(this.didScope);
	    req.setAuthenticationProtocolData(data);

	    DIDAuthenticateResponse res = (DIDAuthenticateResponse) didInfos.getDispatcher().safeDeliver(req);
	    WSHelper.checkResult(res);

	    // check retry counter
	    PINCompareDIDAuthenticateOutputType protoData;
	    protoData = new PINCompareDIDAuthenticateOutputType(res.getAuthenticationProtocolData());
	    BigInteger retryCounter = protoData.getRetryCounter();
	    return retryCounter;
	} catch (ParserConfigurationException ex) {
	    String msg = "Unexpected protocol data received in PIN Compare output.";
	    LOG.error(msg, ex);
	    throw new IllegalStateException(msg);
	}
    }

    @Nullable
    public BigInteger enterPin() throws WSHelper.WSException {
	// pin may be set or it is null
	return enterPin(pin);
    }


    public boolean isCryptoDid() throws WSHelper.WSException {
	return "urn:oid:1.3.162.15480.3.0.25".equals(getProtocol());
    }

    public CryptoMarkerType getGenericCryptoMarker() throws WSHelper.WSException {
	return new CryptoMarkerType(getDID().getDIDMarker());
    }

    public byte[] hash(byte[] data) throws WSHelper.WSException {
	if (! isCryptoDid()) {
	    throw new IllegalStateException("Hash called for a DID which is not a Generic Crypto DID.");
	}

	Hash hashReq = new Hash();
	hashReq.setMessage(data);
	hashReq.setDIDName(didTarget.getDIDName());
	hashReq.setDIDScope(DIDScopeType.LOCAL);
	hashReq.setConnectionHandle(didInfos.getHandle(application));
	HashResponse res = (HashResponse) didInfos.getDispatcher().safeDeliver(hashReq);
	WSHelper.checkResult(res);

	byte[] digest = res.getHash();
	return digest;
    }

    public byte[] sign(byte[] data) throws WSHelper.WSException {
	if (! isCryptoDid()) {
	    throw new IllegalStateException("Sign called for a DID which is not a Generic Crypto DID.");
	}

	Sign sign = new Sign();
	sign.setMessage(data);
	sign.setDIDName(didTarget.getDIDName());
	sign.setDIDScope(DIDScopeType.LOCAL);
	sign.setConnectionHandle(didInfos.getHandle(application));
	SignResponse res = (SignResponse) didInfos.getDispatcher().safeDeliver(sign);
	WSHelper.checkResult(res);

	byte[] sig = res.getSignature();
	return sig;
    }

    public List<DataSetInfo> getRelatedDataSets() throws WSHelper.WSException {
	try {
	    ArrayList<DataSetInfo> result = new ArrayList<>();
	    Set<String> foundDataSets = new HashSet<>();

	    if (isCryptoDid()) {
		CryptoMarkerType m = getGenericCryptoMarker();
		for (CertificateRefType cert : m.getCertificateRefs()) {
		    String datasetName = cert.getDataSetName();
		    //String dsiName = cert.getDSIName();

		    // add if it is not already present in the result list
		    if (! foundDataSets.contains(datasetName)) {
			DataSetInfo ds = didInfos.getDataSetInfo(application, datasetName);
			result.add(ds);
			foundDataSets.add(datasetName);
		    }
		}
	    }

	    return Collections.unmodifiableList(result);
	} catch (NoSuchDataSet ex) {
	    String msg = "DataSet referenced in CIF could not be resolved.";
	    LOG.error(msg, ex);
	    throw WSHelper.createException(WSHelper.makeResultError(ECardConstants.Minor.App.INT_ERROR, msg));
	}
    }

    public synchronized List<X509Certificate> getRelatedCertificateChain() throws WSHelper.WSException,
	    CertificateException, SecurityConditionUnsatisfiable, NoSuchDid {
	if (certs == null) {
	    if (isCryptoDid()) {
		// read certs from card
		boolean allCertsRead = true;
		ArrayList<byte[]> rawCerts = new ArrayList<>();
		for (DataSetInfo dsi : getRelatedDataSets()) {
		    if (dsi.isPinSufficient()) {
			dsi.connectApplication();
			dsi.authenticate();
			byte[] data = dsi.read();
			rawCerts.add(data);
		    } else {
			allCertsRead = false;
			break;
		    }
		}

		// convert certs
		if (allCertsRead) {
		    certs = parseCerts(rawCerts);
		} else {
		    throw new CertificateException("No readable certificates available.");
		}
	    } else {
		certs = Collections.emptyList();
	    }
	}

	return certs;
    }

    private List<X509Certificate> parseCerts(List<byte[]> certsData) throws CertificateException {
	List<Certificate> allCerts = new ArrayList<>();

	for (byte[] nextBlob : certsData) {
	    allCerts.addAll(getCertFactory().generateCertificates(new ByteArrayInputStream(nextBlob)));
	}

	// get first cert and build path
	if (! allCerts.isEmpty()) {
	    return (List<X509Certificate>) getCertFactory().generateCertPath(allCerts).getCertificates();
	} else {
	    return Collections.emptyList();
	}
    }

    public void connectApplication() throws WSHelper.WSException {
	didInfos.connectApplication(application);
    }

    public void authenticateMissing() throws WSHelper.WSException, SecurityConditionUnsatisfiable, NoSuchDid {
	for (DidInfo nextDid : getMissingDidInfos()) {
	    if (nextDid.isPinDid()) {
		nextDid.enterPin(pin);
	    } else {
		throw new SecurityConditionUnsatisfiable("Only PIN DIDs are supported at the moment.");
	    }
	}
    }

}

File: crypto/crypto-common/src/jvmMain/kotlin/org/openecard/crypto/common/sal/did/DidInfo.kt
/****************************************************************************
 * Copyright (C) 2016 ecsec GmbH.
 * All rights reserved.
 * Contact: ecsec GmbH (info@ecsec.de)
 *
 * This file is part of the Open eCard App.
 *
 * GNU General Public License Usage
 * This file may be used under the terms of the GNU General Public
 * License version 3.0 as published by the Free Software Foundation
 * and appearing in the file LICENSE.GPL included in the packaging of
 * this file. Please review the following information to ensure the
 * GNU General Public License version 3.0 requirements will be met:
 * http://www.gnu.org/copyleft/gpl.html.
 *
 * Other Usage
 * Alternatively, this file may be used in accordance with the terms
 * and conditions contained in a signed written agreement between
 * you and ecsec GmbH.
 *
 ***************************************************************************/
package org.openecard.crypto.common.sal.did

import io.github.oshai.kotlinlogging.KotlinLogging
import iso.std.iso_iec._24727.tech.schema.*
import org.openecard.common.ECardConstants
import org.openecard.common.SecurityConditionUnsatisfiable
import org.openecard.common.WSHelper
import org.openecard.common.WSHelper.checkResult
import org.openecard.common.WSHelper.createException
import org.openecard.common.WSHelper.makeResultError
import org.openecard.common.anytype.pin.PINCompareDIDAuthenticateInputType
import org.openecard.common.anytype.pin.PINCompareDIDAuthenticateOutputType
import org.openecard.common.util.ByteUtils
import java.io.ByteArrayInputStream
import java.math.BigInteger
import java.security.cert.Certificate
import java.security.cert.CertificateException
import java.security.cert.CertificateFactory
import java.security.cert.X509Certificate
import java.util.*
import javax.xml.parsers.ParserConfigurationException

private val LOG = KotlinLogging.logger {}

/**
 *
 * @author Tobias Wich
 */
class DidInfo {

	private val didInfos: DidInfos
	private val application: ByteArray
	private val didTarget: TargetNameType
	private var didScope: DIDScopeType?
	private var pin: CharArray?

	@get:Throws(CertificateException::class)
	private val certFactory: CertificateFactory by lazy {
		CertificateFactory.getInstance("X.509")
	}

	internal constructor(didInfos: DidInfos, application: ByteArray, didName: String, pin: CharArray?) {
		this.didInfos = didInfos
		this.application = ByteUtils.clone(application)
		this.didTarget = TargetNameType()
		this.didTarget.setDIDName(didName)
		if (pin != null) {
			this.pin = pin.clone()
		} else {
			this.pin = null
		}

		// we are in the same application, so set it to null and get the real value
		this.didScope = null
		try {
			val didStruct: DIDStructureType? = this.DID
			if (didStruct != null) {
				this.didScope = didStruct.getDIDScope()
			}
		} catch (_: WSHelper.WSException) {
			// too bad ;-)
			LOG.warn { "Failed to retrieve DID qualifier." }
		}
	}

	internal constructor(didInfos: DidInfos, application: ByteArray, didStruct: DIDStructureType, pin: CharArray?) {
		this.didInfos = didInfos
		this.application = application.copyOf()
		this.didTarget = TargetNameType()
		this.didTarget.setDIDName(didStruct.getDIDName())
		this.didScope = didStruct.getDIDScope()
		if (pin != null) {
			this.pin = pin.clone()
		} else {
			this.pin = null
		}
	}

	fun setPin(pin: CharArray?) {
		if (pin != null) {
			this.pin?.fill(' ')
			this.pin = pin.clone()
		} else {
			this.pin = null
		}
	}

	@get:Throws(WSHelper.WSException::class)
	val ACL: AccessControlListType
		get() {
			val req = ACLList()
			req.setConnectionHandle(didInfos.getHandle(application))
			req.setTargetName(didTarget)

			val res = didInfos.dispatcher.safeDeliver(req) as ACLListResponse
			checkResult<ACLListResponse>(res)

			return res.getTargetACL()
		}

	@get:Throws(WSHelper.WSException::class)
	val DID: DIDStructureType
		get() {
			val req = DIDGet()
			req.setConnectionHandle(didInfos.getHandle(application))
			req.setDIDName(this.didName)
			req.setDIDScope(didScope)

			val res = didInfos.dispatcher.safeDeliver(req) as DIDGetResponse
			checkResult<DIDGetResponse>(res)

			return res.getDIDStructure()
		}

	val didName: String by lazy { didTarget.getDIDName() }

	@get:Throws(WSHelper.WSException::class)
	val protocol = this.DID.getDIDMarker().getProtocol()

	@get:Throws(WSHelper.WSException::class)
	val isAuthenticated = this.DID.isAuthenticated

	@get:Throws(
		WSHelper.WSException::class,
		SecurityConditionUnsatisfiable::class
	)
	val missingDids: List<DIDStructureType>
		get() {
			val resolver = ACLResolver(didInfos.dispatcher, didInfos.getHandle(application))
			val missingDids = resolver.getUnsatisfiedDIDs(didTarget, this.ACL.getAccessRule())
			return missingDids
		}

	@get:Throws(
		WSHelper.WSException::class,
		SecurityConditionUnsatisfiable::class,
		NoSuchDid::class
	)
	val missingDidInfos: MutableList<DidInfo>
		get() {
			val result = ArrayList<DidInfo>()
			for (didStruct in this.missingDids) {
				result.add(didInfos.getDidInfo(didStruct.getDIDName()))
			}
			return result
		}

	@get:Throws(WSHelper.WSException::class)
	val isPinSufficient: Boolean
		get() {
			try {
				val missingDids = this.missingDids
				// check if there is anything other than a pin did in the list
				for (missingDid in missingDids) {
					val infoObj: DidInfo = if (missingDid.getDIDScope() == DIDScopeType.GLOBAL) {
						didInfos.getDidInfo(missingDid.getDIDName())
					} else {
						didInfos.getDidInfo(application, missingDid.getDIDName())
					}
					if (!infoObj.isPinDid) {
						return false
					}
				}
				// no PIN DID in missing list
				return true
			} catch (_: SecurityConditionUnsatisfiable) {
				// not satisfiable means pin does not suffice
				return false
			} catch (ex: NoSuchDid) {
				val msg = "DID referenced in CIF could not be resolved."
				LOG.error(ex) { msg }
				throw createException(
					makeResultError(
						ECardConstants.Minor.App.INT_ERROR,
						msg
					)
				)
			}
		}

	@Throws(WSHelper.WSException::class, SecurityConditionUnsatisfiable::class)
	fun needsPin(): Boolean {
		try {
			val missingDids = this.missingDids
			// check if there is a pin did in the list
			for (missingDid in missingDids) {
				val infoObj = if (missingDid.getDIDScope() == DIDScopeType.GLOBAL) {
					didInfos.getDidInfo(missingDid.getDIDName())
				} else {
					didInfos.getDidInfo(application, missingDid.getDIDName())
				}
				if (infoObj.isPinDid) {
					return true
				}
			}
			// no PIN DID in missing list
			return false
		} catch (ex: NoSuchDid) {
			val msg = "DID referenced in CIF could not be resolved."
			LOG.error(ex) { msg }
			throw createException(makeResultError(ECardConstants.Minor.App.INT_ERROR, msg))
		}
	}


	@get:Throws(WSHelper.WSException::class)
	val rawMarker: DIDAbstractMarkerType
		get() = this.DID.getDIDMarker()


	@get:Throws(WSHelper.WSException::class)
	val isPinDid: Boolean
		get() = "urn:oid:1.3.162.15480.3.0.9" == this.protocol

	@get:Throws(WSHelper.WSException::class)
	val pinCompareMarker: PinCompareMarkerType
		get() = PinCompareMarkerType(this.DID.getDIDMarker())

	@JvmOverloads
	@Throws(WSHelper.WSException::class)
	fun enterPin(pin: CharArray? = this.pin): BigInteger {
		check(this.isPinDid) { "Enter PIN called for a DID which is not a PIN DID." }

		try {
			var data = PinCompareDIDAuthenticateInputType()
			data.setProtocol(this.protocol)
			// add PIN
			if (pin != null && pin.isNotEmpty()) {
				val builder = PINCompareDIDAuthenticateInputType(data)
				builder.setPIN(pin)
				data = builder.getAuthDataType()
				builder.setPIN(null)
			}

			val req = DIDAuthenticate()
			req.setConnectionHandle(didInfos.getHandle(application))
			req.setDIDName(didTarget.getDIDName())
			req.setDIDScope(this.didScope)
			req.setAuthenticationProtocolData(data)

			val res = didInfos.dispatcher.safeDeliver(req) as DIDAuthenticateResponse
			checkResult<DIDAuthenticateResponse>(res)

			// check retry counter
			val protoData = PINCompareDIDAuthenticateOutputType(res.getAuthenticationProtocolData())
			val retryCounter = protoData.retryCounter
			return retryCounter
		} catch (ex: ParserConfigurationException) {
			val msg = "Unexpected protocol data received in PIN Compare output."
			LOG.error(ex) { msg }
			throw IllegalStateException(msg)
		}
	}


	@get:Throws(WSHelper.WSException::class)
	val isCryptoDid: Boolean
		get() = "urn:oid:1.3.162.15480.3.0.25" == this.protocol

	@get:Throws(WSHelper.WSException::class)
	val genericCryptoMarker: CryptoMarkerType
		get() = CryptoMarkerType(this.DID.getDIDMarker())

	@Throws(WSHelper.WSException::class)
	fun hash(data: ByteArray): ByteArray {
		check(this.isCryptoDid) { "Hash called for a DID which is not a Generic Crypto DID." }

		val hashReq = Hash()
		hashReq.setMessage(data)
		hashReq.setDIDName(didTarget.getDIDName())
		hashReq.setDIDScope(DIDScopeType.LOCAL)
		hashReq.setConnectionHandle(didInfos.getHandle(application))
		val res = didInfos.dispatcher.safeDeliver(hashReq) as HashResponse
		checkResult<HashResponse>(res)

		val digest = res.getHash()
		return digest
	}

	@Throws(WSHelper.WSException::class)
	fun sign(data: ByteArray): ByteArray {
		check(this.isCryptoDid) { "Sign called for a DID which is not a Generic Crypto DID." }

		val sign = Sign()
		sign.setMessage(data)
		sign.setDIDName(didTarget.getDIDName())
		sign.setDIDScope(DIDScopeType.LOCAL)
		sign.setConnectionHandle(didInfos.getHandle(application))
		val res = didInfos.dispatcher.safeDeliver(sign) as SignResponse
		checkResult<SignResponse>(res)

		val sig = res.getSignature()
		return sig
	}

	@get:Throws(WSHelper.WSException::class)
	val relatedDataSets: List<DataSetInfo>
		get() {
			try {
				val result = ArrayList<DataSetInfo?>()
				val foundDataSets: MutableSet<String> = HashSet()

				if (this.isCryptoDid) {
					val m = this.genericCryptoMarker
					for (cert in m.certificateRefs) {
						val datasetName = cert.getDataSetName()

						//String dsiName = cert.getDSIName();

						// add if it is not already present in the result list
						if (!foundDataSets.contains(datasetName)) {
							val ds = didInfos.getDataSetInfo(datasetName)
							result.add(ds)
							foundDataSets.add(datasetName)
						}
					}
				}

				return Collections.unmodifiableList<DataSetInfo?>(result)
			} catch (ex: NoSuchDataSet) {
				val msg = "DataSet referenced in CIF could not be resolved."
				LOG.error(ex) { msg }
				throw createException(
					makeResultError(
						ECardConstants.Minor.App.INT_ERROR,
						msg
					)
				)
			}
		}

	@get:Throws(
		WSHelper.WSException::class,
		CertificateException::class,
		SecurityConditionUnsatisfiable::class,
		NoSuchDid::class
	)
	@get:Synchronized
	val relatedCertificateChain: List<X509Certificate> by lazy {
		if (this.isCryptoDid) {
			// read certs from card
			var allCertsRead = true
			val rawCerts = ArrayList<ByteArray>()
			for (dsi in this.relatedDataSets) {
				if (dsi.isPinSufficient) {
					dsi.connectApplication()
					dsi.authenticate()
					val data = dsi.read()
					rawCerts.add(data)
				} else {
					allCertsRead = false
					break
				}
			}

			// convert certs
			if (allCertsRead) {
				parseCerts(rawCerts)
			} else {
				throw CertificateException("No readable certificates available.")
			}
		} else {
			listOf()
		}
	}


	@Throws(CertificateException::class)
	private fun parseCerts(certsData: MutableList<ByteArray>): List<X509Certificate> {
		val allCerts: MutableList<Certificate> = ArrayList<Certificate>()

		for (nextBlob in certsData) {
			allCerts.addAll(this.certFactory.generateCertificates(ByteArrayInputStream(nextBlob)))
		}

		// get first cert and build path
		return if (!allCerts.isEmpty()) {
			@Suppress("UNCHECKED_CAST")
			this.certFactory.generateCertPath(allCerts).certificates as List<X509Certificate>
		} else {
			listOf<X509Certificate>()
		}
	}

	@Throws(WSHelper.WSException::class)
	fun connectApplication() {
		didInfos.connectApplication(application)
	}

	@Throws(WSHelper.WSException::class, SecurityConditionUnsatisfiable::class, NoSuchDid::class)
	fun authenticateMissing() {
		for (nextDid in this.missingDidInfos) {
			if (nextDid.isPinDid) {
				nextDid.enterPin(pin)
			} else {
				throw SecurityConditionUnsatisfiable("Only PIN DIDs are supported at the moment.")
			}
		}
	}

}
